### DISCUSSION TITLE
Building for debugging?

### DISCUSSION THREAD

**Discussion by roccomoretti** [molstar/molstar#855] [General]
I'm trying to understand the internals of Mol* better, and I've been trying to use the javascript debugging features of Chrome to trace through how various things happen. 

The problem I'm running into is that using the debugger on Mol* results in tracing through the generated javascript code, rather than the original typescript code. (And unfortunately, the generated javascript is less-than-user-friendly, so attempting to trace it doesn't really work.) As I understand it, it should be possible to generate source maps which can allow Chrome to display the original typescript source for debugging, rather than the generated javascript. However, I haven't been able to figure out how to tweak the build settings to get something which actually works.

How do Mol* developers build Mol* for use with a debugger? Are there settings which would allow typescript source maps to be generated?

**Comment by dsehnal** [comment]
You can try `npm run "watch-viewer-debug`, that should generate the source maps as expected. If it doesn't, maybe you can figure out the correct webpack settings in `webpack.config.viewer.debug.js`?

**Comment by roccomoretti** [comment]
It doesn't look like watch-viewer-debug generates source maps. From what I can tell, `devtool: 'eval'` doesn't generate source maps (though I admit I might be doing something wrong), but if I change it to `devtool: 'source-map'`, I do see source maps. Unfortunately, that's still only the generated javascript sources, rather than the original typescript sources. 

Editing `tsconfig.commonjs.json` (and  `tsconfig.json`) such that the lines read `"sourceMap": true,`  and rebuilding does seem to generate *.map files in the lib/ directory, but it doesn't look like those maps are used for the webpack-generated bundle. In Chrome I'm still seeing the generated javascript, rather than the original typescript files.

[This StackOverflow answer](https://stackoverflow.com/a/62222479) indicates that there's some settings regarding source-map-loader which are needed to enable webpack to load the pre-generated typescript-javascript source maps, and including those changes along with the others seems to work.

I don't know if that's the best way to do it, though.

**Reply by dsehnal** [reply]
Well, if you can figure out the correct set of incantations needed to make this work, I am sure it would be a nice pull request to contribute to the project.
